### PR TITLE
Mozilla VPN landing page

### PR DESCRIPTION
**Pull request by alexgibson**
## Description
Migrates Mozilla VPN [landing page](https://vpn.mozilla.org/)  and [invite page](https://vpn.mozilla.org/vpn/invite) to www.

- http://localhost:8000/en-US/products/vpn/
- http://localhost:8000/en-US/products/vpn/invite/

**Note:** strings are hard-coded for en-US only right now. We'll be localizing the page pending some design updates supporting the upcoming EU VPN launch.

## Issue / Bugzilla link
https://github.com/mozilla/bedrock/issues/9855

## Testing
**Demo:**
- https://www-demo1.allizom.org/en-US/products/vpn/
- https://www-demo1.allizom.org/en-US/products/vpn/invite/

**Successful test run:**
https://gitlab.com/mozmeao/www-config/-/pipelines/252089521

**Notes:**
You can mock geolocation by passing a `geo` parameter to the landing page URL:
- [ ] VPN available: http://localhost:8000/en-US/products/vpn/?geo=us
- [ ] VPN not available: http://localhost:8000/en-US/products/vpn/?geo=de

### CODE REVIEW

File: bedrock/products/tests/test_views.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+        return json.loads(resp.content)
+
+    def test_form_success(self):
+        resp_data = self._request({
+            'newsletters': 'guardian-vpn-waitlist',
+            'email': 'test@example.com',
+            'country': 'us',
+            'privacy': True,
+            'fmt': 'H',
+            'lang': ''
```
Comment (alexgibson) [inline bedrock/products/tests/test_views.py:36]: @pmac thanks again for helping out yesterday. I tried to put together some basic tests for the new view earlier this morning, and what I have here seems to be working ok. But one strange thing I couldn't quite figure out is why the tests here won't accept any value for `lang`, other than an empty string? If I try and pass `en`, which is the default used in the page, then the tests say it's not an accepted value. Any idea where I'm going wrong here? 

(also: not urgent, thanks)

Comment (pmac) [reply bedrock/products/tests/test_views.py:36]: I'm guessing it has to do with how the form populates its language choices. It uses basket and product-details data to provide the choices list. If during the tests these data aren't available I'm guessing the list of choices will be empty. Are you running these locally? If so maybe run `bin/run-db-download.py --force` to get a fresh DB and see if that makes a difference. When I do that and then run the following I do get `en` as a choice:

```python
>>> from bedrock.newsletter.forms import get_lang_choices
>>> get_lang_choices('guardian-vpn-waitlist')
[['en', 'English']]
```

Comment (pmac) [reply bedrock/products/tests/test_views.py:36]: Also nice work on these tests! 🥇 for tests!!

Comment (alexgibson) [reply bedrock/products/tests/test_views.py:36]: Ah, looks like the tests don't have access to this when they run. I solved it with:

```
@patch('bedrock.newsletter.forms.get_lang_choices', lambda *x: [['en', 'English']])
```

Thanks!

**Comment by alexgibson** [comment]
Leaving this PR as darft until stakeholder approval / QA is final, but this is now ready for code review.

File: media/css/products/vpn/landing.scss [hunk 2]
```diff
@@ -0,0 +1,424 @@
...
+    @include text-title-md;
+    margin-bottom: $spacing-2xl;
+    text-align: center;
+}
+
+.vpn-faq-item-title {
+    background: $color-marketing-gray-20;
+    box-shadow: 8px 8px 0 $color-black;
+    margin-bottom: $spacing-lg;
+    padding: $spacing-lg;
```
Comment (maxxcrawford) [inline media/css/products/vpn/landing.scss:358]: The spacing in the open item is much larger than vpn.moz. It's about ~40px. I recommend loosing the bottom margin here. The padding around the `.vpn-faq-item-title` and `.vpn-faq-item-detail` is enough. 

```suggestion
    padding: $spacing-lg;
```

vpn.moz.org: 
image

moz.org/vpn: 
image

File: bedrock/products/templatetags/misc.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+    Examples
+    ========
+
+    In Template
+    -----------
+
+        {{ vpn_sign_in_link(entrypoint='mozilla.org', button_text='Sign In') }}
+    """
+    product_url = 'https://vpn.mozilla.org/oauth/init/'
```
Comment (maxxcrawford) [inline bedrock/products/templatetags/misc.py:49]: Just wanting to confirm that we plan on keeping the `vpn.mozilla.org` URL active for auth/services? I wouldn't want a redirect to break these functions. 

/cc @pmclanahan

File: bedrock/products/tests/test_helper_misc.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+    def test_vpn_subscribe_link(self):
+        """Should return expected markup"""
+        markup = self._render(entrypoint='mozilla.org', link_text='Try Mozilla VPN',
+                              class_name='vpn-button try js-try-vpn',
+                              optional_parameters={'utm_campaign': 'vpn-landing-page'},
+                              optional_attributes={'data-cta-text': 'Try Mozilla VPN', 'data-cta-type':
+                                                   'fxa-vpn', 'data-cta-position': 'primary'})
+        expected = (
+            u'<a href="https://vpn.mozilla.org/r/vpn/subscribe/?entrypoint=mozilla.org&form_type=button'
```
Comment (maxxcrawford) [inline bedrock/products/tests/test_helper_misc.py:42]: In other spots where we set entrypoint params for FxA, we use `entrypoint=mozilla.org-{service/UID}`. Should we add a `-vpn`, etc. here? 

From vpn.moz.org, they use `entrypoint=www.mozilla.org-vpn-product-page`. It may make sense to use the same ID here for joining Amplitude analytics. 

```suggestion
            u'<a href="https://vpn.mozilla.org/r/vpn/subscribe/?entrypoint=mozilla.org-vpn&form_type=button'
```

Other references: 
image

File: bedrock/products/tests/test_helper_misc.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+                      {'request': req})
+
+    def test_vpn_sign_in_link(self):
+        """Should return expected markup"""
+        markup = self._render(entrypoint='mozilla.org', link_text='Sign In', class_name='mzp-c-cta-link',
+                              optional_parameters={'utm_campaign': 'vpn-landing-page'},
+                              optional_attributes={'data-cta-text': 'Sign In', 'data-cta-type':
+                                                   'fxa-vpn', 'data-cta-position': 'navigation'})
+        expected = (
+            u'<a href="https://vpn.mozilla.org/oauth/init/?entrypoint=mozilla.org&form_type=button'
```
Comment (maxxcrawford) [inline bedrock/products/tests/test_helper_misc.py:68]: Same `entrypoint=mozilla.org` callout here.

File: bedrock/newsletter/views.py [hunk 6]
```diff
@@ -663,7 +663,7 @@ def newsletter_subscribe(request):
                     errors.extend(form.errors[fieldname])
 
         # form error messages may contain unsanitized user input
-        errors = list(map(escape, errors))
+        errors = [escape(e) for e in errors]
```
Comment (maxxcrawford) [inline bedrock/newsletter/views.py:666]: Not a comment: What/why this change?

File: bedrock/settings/base.py [hunk 7]
```diff
@@ -1313,3 +1314,34 @@ def get_default_gateway_linux():
...
+  'PR', #Puerto Rico
+  'UM', #United States Minor Outlying Islands
+  'US', #United States of America
+  'VI', #Virgin Islands (U.S.)
+]
+
+# Mozilla VPN subscription prices
+VPN_PRICE_MONTHLY = {
+    'US': {
+        'monthly': 'US$4.99'
```
Comment (maxxcrawford) [inline bedrock/settings/base.py:1345]: TIL: 
image

_I was bugging about the lack of a space between `US` and `$` and wanted to confirm this is noted correctly!_

File: media/js/products/vpn/invite.js [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+
+            // create error list container if it does not exist.
+            if (!newsletterErrors) {
+                newsletterErrors = document.createElement('div');
+                newsletterErrors.id = 'newsletter-errors';
+                newsletterErrors.className = 'mzp-c-form-errors';
+                newsletterContent.insertBefore(newsletterErrors, newsletterContent.firstChild);
+            }
+
+            var errorList = document.createElement('ul');
```
Comment (maxxcrawford) [inline media/js/products/vpn/invite.js:48]: nit: Add a class name of `mzp-u-list-styled` to this `ul`. This will give it the same bulleted look as demoed on the [Protocol site](https://protocol.mozilla.org/patterns/atoms/forms.html#error-list). 

Protocol site: 
image


moz.org/vpn: 
image

**Review by maxxcrawford — CHANGES_REQUESTED** [review]
Overall, this all looks great and REALLY matches the current site well. Great job translating that, plus all the FxA wizardy to wire this up. 

I found a few issues across the board (visual, copy, params, etc.) but nothing critical. 

Will this get tested officially by the QA/PI team before launching? 

Inital browser testing in: 
- Firefox Desktop (latest) 
- IE 11 
- Edge 88
- iOS (latest)

_(Also this PR is still in draft mode!)_

**Comment by maxxcrawford** [comment]
> Note: strings are hard-coded for en-US only right now. We'll be localizing the page pending some design updates supporting the upcoming EU VPN launch. 

Do we have an issue filed for this?

File: bedrock/products/templates/products/vpn/landing.html [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+{% extends "products/vpn/base.html" %}
+
+{% block page_title_full %}Mozilla VPN: Protect Your Entire Device{% endblock %}
+{% block page_title_suffix %}{% endblock %}
+
+{% block page_desc %}Use the Mozilla VPN for full-device protection for all apps. With servers in 30+ countries, you can connect to anywhere, from anywhere.{% endblock %}
+
+{% block body_id %}mozilla-vpn-landing{% endblock %}
+
+{% set _utm_source = 'www.mozilla.org-vpn-product-page' %}
```
[the remark this answers is not in the document]
Comment (alexgibson) [reply bedrock/products/templates/products/vpn/landing.html:16]: This is what has been defined [in the spec](https://docs.google.com/document/d/1uXeiVno6xUGttsjqfN_9YXzIdV66I9F2iFiQ8_XHKIg/edit?usp=sharing), so we can distinguish that it's from www specifically (as opposed to *.mozilla.org, such as vpn.mozilla.org)

File: bedrock/products/templates/products/vpn/landing.html [hunk 10]
```diff
@@ -0,0 +1,240 @@
...
+      {{ join_waitlist_button(link_text='Join the Waitlist', is_conditional=False, position='secondary') }}
+      <p class="availability-copy">We currently offer Mozilla VPN in the US, the UK, Canada, New Zealand, Singapore and Malaysia.</p>
+    {% endcall %}
+
+    {% call vpn_content_well(
+      title='About our trusted partner',
+      image_url='img/products/vpn/landing/vpn-cntwell-06.png',
+      class='vpn-content-well-left-half'
+    ) %}
+      <p>The Mozilla VPN runs on a global network of servers powered by <a href="https://mullvad.net/help/no-logging-data-policy/">Mullvad</a> using the <a href="https://mullvad.net/help/why-wireguard/">WireGuard</a>® protocol. Mullvad puts your privacy first and does not keep logs of any kind.</p>
```
[the remark this answers is not in the document]
Comment (alexgibson) [reply bedrock/products/templates/products/vpn/landing.html:171]: The landing page is now build on Protocol (as far as foundations go), so it's OK to inherit base styles such as dewfault link styles.

[the remark this answers is not in the document]
Comment (alexgibson) [reply bedrock/products/templates/products/vpn/landing.html:171]: Pages render using UTF-8, so there shouldn't be a need to use an HTMl entity code 👍

File: media/css/products/vpn/landing.scss (hunk 2 above)
Comment (alexgibson) [reply media/css/products/vpn/landing.scss:358]: Good call 👍

File: bedrock/products/templatetags/misc.py (hunk 3 above)
Comment (alexgibson) [reply bedrock/products/templatetags/misc.py:49]: www.mozilla.org will host only the VPN product landing page and waitlist pages. The rest of the functionality will remain at vpn.mozilla.org (such as authentication, purchasing, downloads)

File: bedrock/products/tests/test_helper_misc.py (hunk 4 above)
Comment (alexgibson) [reply bedrock/products/tests/test_helper_misc.py:42]: On the landing page we're using `www.mozilla.org-vpn-product-page` for entrypoint. On other pages we'll use things like `www.mozilla.org-vpn-faq-page`, `www.mozilla.org-vpn-comparison-page`. I'll update this test example to follow suit 👍

File: bedrock/newsletter/views.py (hunk 6 above)
Comment (alexgibson) [reply bedrock/newsletter/views.py:666]: I think @pmac added this in his commit, but it seems like it's to make sure we're escaping error messages from the server correctly?

File: bedrock/settings/base.py (hunk 7 above)
Comment (alexgibson) [reply bedrock/settings/base.py:1345]: Yeah it reads strange to me too, but for now it's as correct as can be. In the near future we'll be adding more currencies, so I'd like to revisit this wording when we get to localization.

File: media/js/products/vpn/invite.js (hunk 8 above)
Comment (alexgibson) [reply media/js/products/vpn/invite.js:48]: Good catch!

**Comment by alexgibson** [comment]
> Note: strings are hard-coded for en-US only right now. We'll be localizing the page pending some design updates supporting the upcoming EU VPN launch.

> Do we have an issue filed for this?

Not yet, but coming soon.

**Comment by alexgibson** [comment]
> Will this get tested officially by the QA/PI team before launching?

QA is underway with the product / data science team (hence why this PR is still draft) 👍

**Comment by alexgibson** [comment]
Thanks for the review @maxxcrawford! I added a followup commit with some suggested fixes 👍

File: bedrock/newsletter/views.py (hunk 6 above)
Comment (pmac) [reply bedrock/newsletter/views.py:666]: I'm sure that was me. It's just more clear what is happening and it's supposed to be a bit quicker and more efficient. It's doing the same thing.

**Review by maxxcrawford — APPROVED** [review]
